// _umi-build-dev@1.2.6@umi-build-dev/lib/plugins/commands/generate/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _path = require("path");

var _assert = _interopRequireDefault(require("assert"));

var _chalk = _interopRequireDefault(require("chalk"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(api) {
  const generators = api.service.generators,
        log = api.log;

  function generate(args = {}) {
    try {
      const name = args._[0];
      (0, _assert.default)(name, `run ${_chalk.default.cyan.underline('umi help generate')} to checkout the usage`);
      (0, _assert.default)(generators[name], `Generator ${_chalk.default.cyan.underline(name)} not found`);
      const _generators$name = generators[name],
            Generator = _generators$name.Generator,
            resolved = _generators$name.resolved;
      const generator = new Generator(process.argv.slice(4), {
        env: {
          cwd: api.cwd
        },
        resolved: resolved || __dirname
      });
      generator.run().then(() => {
        log.success('');
      }).catch(e => {
        log.error(e);
      });
    } catch (e) {
      log.error(`Generate failed, ${e.message}`);
    }
  }

  const details = `
Examples: 

  ${_chalk.default.gray('# generate page users')}
  umi generate page users
  
  ${_chalk.default.gray('# g is the alias for generate')}
  umi g page index
  `.trim();
  api.registerCommand('g', {
    description: 'generate code snippets quickly (alias for generate)',
    usage: 'umi g name args',
    details
  }, generate);
  api.registerCommand('generate', {
    description: 'generate code snippets quickly',
    usage: 'umi generate name args',
    details
  }, generate);
  api.registerGenerator('page', {
    Generator: require('./page').default(api),
    resolved: (0, _path.join)(__dirname, './page')
  });
}
