// _umi-build-dev@1.2.6@umi-build-dev/lib/plugins/mock/createMockMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getMockMiddleware;

var _fs = require("fs");

var _path = require("path");

var _bodyParser = _interopRequireDefault(require("body-parser"));

var _glob = _interopRequireDefault(require("glob"));

var _assert = _interopRequireDefault(require("assert"));

var _chokidar = _interopRequireDefault(require("chokidar"));

var _pathToRegexp = _interopRequireDefault(require("path-to-regexp"));

var _signale = _interopRequireDefault(require("signale"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const VALID_METHODS = ['get', 'post', 'put', 'patch', 'delete'];
const BODY_PARSED_METHODS = ['post', 'put', 'patch'];

function getMockMiddleware(api, errors) {
  const debug = api.debug;
  const cwd = api.service.cwd;
  const absMockPath = (0, _path.join)(cwd, 'mock');
  const absConfigPath = (0, _path.join)(cwd, '.umirc.mock.js');
  api.addBabelRegister([absMockPath, absConfigPath]);
  let mockData = getConfig();
  watch();

  function watch() {
    if (process.env.WATCH_FILES === 'none') return;

    const watcher = _chokidar.default.watch([absConfigPath, absMockPath], {
      ignoreInitial: true
    });

    watcher.on('all', (event, file) => {
      debug(`[${event}] ${file}, reload mock data`);
      mockData = getConfig();

      if (!errors.length) {
        _signale.default.success(`Mock file parse success`);
      }
    });
  }

  function getConfig() {
    // Clear errors
    errors.splice(0, errors.length);
    cleanRequireCache();
    let ret = {};

    if ((0, _fs.existsSync)(absConfigPath)) {
      debug(`load mock data from ${absConfigPath}`);
      ret = require(absConfigPath); // eslint-disable-line
    } else {
      const mockFiles = _glob.default.sync('**/*.js', {
        cwd: absMockPath
      });

      debug(`load mock data from ${absMockPath}, including files ${JSON.stringify(mockFiles)}`);

      try {
        ret = mockFiles.reduce((memo, mockFile) => {
          const m = require((0, _path.join)(absMockPath, mockFile)); // eslint-disable-line


          memo = _objectSpread({}, memo, m.default || m);
          return memo;
        }, {});
      } catch (e) {
        errors.push(e);

        _signale.default.error(`Mock file parse failed`);

        console.error(e.message);
      }
    }

    return normalizeConfig(ret);
  }

  function parseKey(key) {
    let method = 'get';
    let path = key;

    if (key.indexOf(' ') > -1) {
      const splited = key.split(' ');
      method = splited[0].toLowerCase();
      path = splited[1]; // eslint-disable-line
    }

    (0, _assert.default)(VALID_METHODS.includes(method), `Invalid method ${method} for path ${path}, please check your mock files.`);
    return {
      method,
      path
    };
  }

  function createHandler(method, path, handler) {
    return function (req, res, next) {
      if (BODY_PARSED_METHODS.includes(method)) {
        _bodyParser.default.json({
          limit: '5mb',
          strict: false
        })(req, res, () => {
          _bodyParser.default.urlencoded({
            limit: '5mb',
            extended: true
          })(req, res, () => {
            sendData();
          });
        });
      } else {
        sendData();
      }

      function sendData() {
        if (typeof handler === 'function') {
          handler(req, res, next);
        } else {
          res.json(handler);
        }
      }
    };
  }

  function normalizeConfig(config) {
    return Object.keys(config).reduce((memo, key) => {
      const handler = config[key];
      const type = typeof handler;
      (0, _assert.default)(type === 'function' || type === 'object', `mock value of ${key} should be function or object, but got ${type}`);

      const _parseKey = parseKey(key),
            method = _parseKey.method,
            path = _parseKey.path;

      const keys = [];
      const re = (0, _pathToRegexp.default)(path, keys);
      memo.push({
        method,
        path,
        re,
        keys,
        handler: createHandler(method, path, handler)
      });
      return memo;
    }, []);
  }

  function cleanRequireCache() {
    Object.keys(require.cache).forEach(file => {
      if (file === absConfigPath || file.indexOf(absMockPath) > -1) {
        delete require.cache[file];
      }
    });
  }

  function matchMock(req) {
    const exceptPath = req.path;
    const exceptMethod = req.method.toLowerCase();
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = mockData[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        const mock = _step.value;
        const method = mock.method,
              re = mock.re,
              keys = mock.keys;

        if (method === exceptMethod) {
          const match = re.exec(req.path);

          if (match) {
            const params = {};

            for (let i = 1; i < match.length; i = i + 1) {
              const key = keys[i - 1];
              const prop = key.name;
              const val = decodeParam(match[i]);

              if (val !== undefined || !hasOwnProperty.call(params, prop)) {
                params[prop] = val;
              }
            }

            req.params = params;
            return mock;
          }
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    function decodeParam(val) {
      if (typeof val !== 'string' || val.length === 0) {
        return val;
      }

      try {
        return decodeURIComponent(val);
      } catch (err) {
        if (err instanceof URIError) {
          err.message = `Failed to decode param ' ${val} '`;
          err.status = err.statusCode = 400;
        }

        throw err;
      }
    }

    return mockData.filter(({
      method,
      re
    }) => {
      return method === exceptMethod && re.test(exceptPath);
    })[0];
  }

  return function UMI_MOCK(req, res, next) {
    const match = matchMock(req);

    if (match) {
      debug(`mock matched: [${match.method}] ${match.path}`);
      return match.handler(req, res, next);
    } else {
      return next();
    }
  };
}
