// _umi-build-dev@1.2.6@umi-build-dev/lib/getPaths.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _path = require("path");

var _fs = require("fs");

function test(path) {
  return (0, _fs.existsSync)(path) && (0, _fs.statSync)(path).isDirectory();
}

function template(path) {
  return (0, _path.join)(__dirname, '../template', path);
}

function _default(service) {
  const cwd = service.cwd,
        config = service.config;
  const outputPath = config.outputPath || './dist';
  let pagesPath = 'pages';

  if (process.env.PAGES_PATH) {
    pagesPath = process.env.PAGES_PATH;
  }

  if (test((0, _path.join)(cwd, 'src/page'))) {
    pagesPath = 'src/page';
  }

  if (test((0, _path.join)(cwd, 'src/pages'))) {
    pagesPath = 'src/pages';
  }

  if (test((0, _path.join)(cwd, 'page'))) {
    pagesPath = 'page';
  }

  const absPagesPath = (0, _path.join)(cwd, pagesPath);
  const absSrcPath = (0, _path.join)(absPagesPath, '../');
  const envAffix = process.env.NODE_ENV === 'development' ? '' : `-production`;
  const tmpDirPath = `${pagesPath}/.umi${envAffix}`;
  const absTmpDirPath = (0, _path.join)(cwd, tmpDirPath);
  return {
    cwd,
    outputPath,
    absOutputPath: (0, _path.join)(cwd, outputPath),
    absNodeModulesPath: (0, _path.join)(cwd, 'node_modules'),
    pagesPath,
    absPagesPath,
    absSrcPath,
    tmpDirPath,
    absTmpDirPath,
    absRouterJSPath: (0, _path.join)(absTmpDirPath, 'router.js'),
    absLibraryJSPath: (0, _path.join)(absTmpDirPath, 'umi.js'),
    absRegisterSWJSPath: (0, _path.join)(absTmpDirPath, 'registerServiceWorker.js'),
    absPageDocumentPath: (0, _path.join)(absPagesPath, 'document.ejs'),
    defaultEntryTplPath: template('entry.js.tpl'),
    defaultRouterTplPath: template('router.js.tpl'),
    defaultHistoryTplPath: template('history.js.tpl'),
    defaultRegisterSWTplPath: template('registerServiceWorker.js'),
    defaultDocumentPath: template('document.ejs')
  };
}
